### PR TITLE
add CHELSA V2

### PR DESCRIPTION
**Pull request by tiemvanderdeure**
This package currently downloads CHELSA data from V1. 

CHELSA V2 has a bunch of additional variables, so I want to make those accessible from this package.

I think version 2 should become the default, but we need to still allow for version 1. The easiest way to do that would be to have version as a keyword argument for the `CHELSA` type

To-do list
- Version argument?
- [Bioclim+ variables](https://chelsa-climate.org/exchelsa-extended-bioclim/)
- Version 2 for future climate

### CODE REVIEW

**Comment by rafaqz** [comment]
Sure Chelsa version type parameter makes sense, 2 is assumed if not specified

**Comment by rafaqz** [comment]
Re test failures: the Aqua.jl keywords have changed. We'll need to remove the toml formatting part and also probably specify the Aqua.jl version in Project.toml

File: src/types.jl [hunk 1]
```diff
@@ -65,6 +65,38 @@ bioclim_key(k::Integer) = keys(bioclim_lookup)[k]
...
+Currently implemented for CHELSA as `CHELSA{BioClim}` and `CHELSA{Future{BioClim, args..}}`,
+specifying layer names as `Symbol`s.
+
+See the [`getraster`](@ref) docs for implementation details.
+"""
+struct BioClimPlus <: RasterDataSet end
+
+layerkeys(T::Type{BioClimPlus}) = vcat(collect(layerkeys(BioClim)), bioclimplus_layerkeys)
+
+const _bioclimplus_monthly = (:hurs, :clt, :sfcWind, :vpd, :rsds, :pet, :cmi, :swb)
```
Comment (rafaqz) [inline src/types.jl]: Good to make constants UPPER_CASE so we know where they come from when you use them

File: src/chelsa/shared.jl [hunk 2]
```diff
@@ -10,4 +10,13 @@ and the [`getraster`](@ref) docs for implementation details.
...
 
 rasterpath(::Type{CHELSA}) = joinpath(rasterpath(), "CHELSA")
-rasterurl(::Type{CHELSA}) = URI(scheme="https", host="os.zhdk.cloud.switch.ch", path="/envicloud/chelsa/")
+rasterurl(::Type{CHELSA}, ::Val{2}) = URI(scheme="https", host="os.zhdk.cloud.switch.ch", path="/envicloud/chelsa/chelsa_V2/GLOBAL/")
+rasterurl(::Type{CHELSA}, ::Val{1}) = URI(scheme="https", host="os.zhdk.cloud.switch.ch", path="/envicloud/chelsa/chelsa_V1/")
+
+latest_patch(c::Type{<:CHELSA}, ::Val{1}) = 2
+latest_patch(c::Type{<:CHELSA}, ::Val{2}) = 1
+
+CHELSA_KEYWORDS = """
```
Comment (rafaqz) [inline src/chelsa/shared.jl]:
```suggestion
const CHELSA_KEYWORDS = """
```

File: src/chelsa/bioclim.jl [hunk 3]
```diff
@@ -11,23 +12,57 @@ Download [`CHELSA`](@ref) [`BioClim`](@ref) data from [chelsa-climate.org](https
...
-rasterurl(::Type{CHELSA{BioClim}}) = joinpath(rasterurl(CHELSA), "chelsa_V2/GLOBAL/climatologies/1981-2010/bio/")
-rasterurl(T::Type{CHELSA{BioClim}}, layer::Integer) = joinpath(rasterurl(T), rastername(T, layer))
+### Bioclim+
+layers(::Type{CHELSA{BioClimPlus}}) = layers(BioClimPlus)
+
+getraster(
+    T::Type{CHELSA{BioClimPlus}}, 
+    layer::Union{Tuple,Int,Symbol}; 
+    version::Int = 2, 
+    patch::Int = latest_patch(T, Val(version))) = _getraster(T, layer, Val(version), patch)
```
Comment (rafaqz) [inline src/chelsa/bioclim.jl]: Try not to lift runtime values to the type domain if we don't need to. Cant we just use a conditional on `version` or interpolate it into the url?

Comment (tiemvanderdeure) [reply src/chelsa/bioclim.jl]: The url is a bit different between the versions, but we can just use some if-else.

File: src/types.jl (hunk 1 above)
Comment (tiemvanderdeure) [reply src/types.jl]: See the point. `bioclim_lookup` is lowercase, so I tried to keep that consistent.

Comment (rafaqz) [reply src/types.jl]: Should be uppercase too if its a constant ;)

File: src/chelsa/bioclim.jl (hunk 3 above)
Comment (rafaqz) [reply src/chelsa/bioclim.jl]: We could even put the version in the CHELSA type then? And no version is the default? 

Introducing keywords means Rasters.jl has to forward them.

Comment (tiemvanderdeure) [reply src/chelsa/bioclim.jl]: Is it a problem to forward the keywords? If we just add `version` and `patch` to `getraster_keywords`, that should just work.

Comment (rafaqz) [reply src/chelsa/bioclim.jl]: I forgot we had added `getraster_keywords` for handling this. It was much more annoying before that.

**Comment by tiemvanderdeure** [comment]
Just need to add some tests and this is good to go.

For future climate, I haven't added a version keyword because CMIP6 is always version 2.

The bioclim variables are in the exact same path as the bioclim+ variables, so we could consider making it so that `getraster(CHELSA{BioClim})` calls `getraster(CHELSA{BioClimPlus})`, and that will save us some lines of code.

**Comment by tiemvanderdeure** [comment]
@rafaqz could you review this and help me troubleshoot these test failures?

**Comment by rafaqz** [comment]
Ok so Aqua.jl dosen't  do toml formatting anymore, yoy can delete that line from the tests.

**Comment by rafaqz** [comment]
Aqua is in the runtest.jl file, near the top. While youre at it maybe add a line there to `include("chelsa_climate.jl")` ;)

**Comment by tiemvanderdeure** [comment]
@rafaqz It now fails because of codecov. Could you take a look?

**Comment by rafaqz** [comment]
This looks good to go. Do you want to also update the readme table so CHELSA includes plus and climate?

(We can ignore that error its just codecov)

**Comment by tiemvanderdeure** [comment]
FYI @rafaqz - precompilation gives this output. I can't figure out which line triggers this (nothing to do with this PR I think - it's the same on master)

```
Precompiling RasterDataSources
        Info Given RasterDataSources was explicitly requested, output will be shown live
┌ Warning: Replacing docs for `RasterDataSources.getraster :: Union{}` in module `RasterDataSources`   
└ @ Base.Docs docs\Docs.jl:243
```

**Comment by rafaqz** [comment]
Yeah would be nice to get a line number for that...

Fixed in #65

**Comment by rafaqz** [comment]
How can WorldClim downloads be this unreliable

**Comment by tiemvanderdeure** [comment]
Are you kidding me, worldclim is down again!! I think this PR is done - can you re-trigger tests manually when worldclim is up again @rafaqz?

**Comment by rafaqz** [comment]
Yeah, maybe we need to cut any testing dependency on them or make it optional based on if it loads at all... Or maybe we can use GitHub actions cache somehow to cache the downloads now 

It's so weird for such a widely used dataset.

**Comment by tiemvanderdeure** [comment]
@rafaqz I think this PR is good to go - the tests just failed because worldclim was down. Can you manually re-run the tests? And maybe give the code a final glance?

**Comment by rafaqz** [comment]
Thanks, great to have this :)

**Pull request merged**